### PR TITLE
Document the pre-commit.ci bot

### LINKED ISSUE #13320: Start using pre-commit ci app
### Description

Astropy should consider using the pre-commit ci app: https://github.com/apps/pre-commit-ci, instead of its current `Code style checks` github action for PRs.

### Additional context

This will provide several benefits including:

- Automatic fixing of code style issues (current action uses pre-commit to enforce code style). Which should help ease the burden of code style issues on contributors. Note that this is a feature we can choose to disable globally, but might make user contributions easier since contributors would not need to install ``pre-commit``, ``isort``, ``black``, etc if they don't want to and can just pull fixed PRs from GH.
- Automated PRs to update our pre-commit configuration.

See https://pre-commit.ci for details on the setup options.

See [dev discussion](https://docs.google.com/document/d/15JSFh3OMF9Iz6ov3q_xxGO_BL8hRnuse4IMUrqEIvcg/edit#bookmark=id.wcqlrvx3kacw) for more details.

### Blocked by

* https://github.com/pre-commit-ci/issues/issues/134

### PR DESCRIPTION
**Pull request by WilliamJamieson**
### Description

This pull request documents how to interact with and use the pre-commit.ci bot and removes the old codestyle action (this will be taken over by the bot).

NOTE: this PR should not be merged until #13552 is merged and the [pre-commit.ci](https://pre-commit.ci/) bot has been enabled in astropy.

Fixes #13320

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: docs/development/workflow/maintainer_workflow.rst
Comment (taldcroft) [inline docs/development/workflow/maintainer_workflow.rst:112]: FWIW trying to soften the language here...
```suggestion
with maintaining the astropy coding style and fixing code style
issues. The bot makes use of the pre-commit hook described in detail in :ref:`pre-commit`.
```

File: docs/development/workflow/maintainer_workflow.rst
Comment (taldcroft) [inline docs/development/workflow/maintainer_workflow.rst:115]: Awesome!

**Review by taldcroft — COMMENTED** [review]
This looks great to me.

**Comment by taldcroft** [comment]
BTW what bot is actually responsible for filling the checkbox in the description? I didn't see anything about that in the precommit.ci docs, but maybe I missed it.

**Comment by WilliamJamieson** [comment]
> BTW what bot is actually responsible for filling the checkbox in the description? I didn't see anything about that in the precommit.ci docs, but maybe I missed it.

We decided we didn't want the bot automatically autofixing PRs so that contributors wouldn't get possible merge conflicts if it fixes their code and they don't do a pull before they make additional changes. Instead we decided that it would be best to ignore the style errors that the bot is detecting until an appropriate point (such as right before a merge). At that point the maintainer handling the PR could make the "magic comment" and fix the PR.

I added the check box so that a maintainer could mark it to remind them that they needed to do this prior to hitting the merge button. I am open to suggestions/changes to this.

**Comment by taldcroft** [comment]
@WilliamJamieson - oops. I misinterpreted "checks section" to mean that it was somehow autofilling the checklist section in the PR description up top. Now I see that you mean the CI checks area. So all good!

**Comment by pllim** [comment]
@Cadair , I don't see Giles bot on this one either. Is it not working still?

**Comment by Cadair** [comment]
Yeah there's some issue apparently specific to the astropy config. I will debug when I find the time.

File: .github/workflows/ci_workflows.yml
Comment (nstarman) [inline .github/workflows/ci_workflows.yml:55]: Just checking, do the more extensive tests run if pre-commit fails? If not, then we need some conditional logic...

Comment (WilliamJamieson) [reply .github/workflows/ci_workflows.yml:55]: Should we turn this into a label triggered CI?

Comment (nstarman) [reply .github/workflows/ci_workflows.yml:55]: I mean, currently the code style check runs quickly relative to the full tests. If the style check fails the other tests stop because of ``fail-fast: true``. Switching pre-commit to the CI bot means these tests don't fail-fast, right? If they do, excellent, if not, then we should make these tests start only if the pre-commit tests succeeds.

Comment (WilliamJamieson) [reply .github/workflows/ci_workflows.yml:55]: With the pre-commit.ci bot, I don't see a big need to have all the CI grind to a halt due to failing codestyle checks. The bot can be commanded to autofix the majority of issues that cause codestyle failures anyways. The plan is to have the bot fix any issues right before we merge the PR anyways.

Comment (WilliamJamieson) [reply .github/workflows/ci_workflows.yml:55]: Note that we should also mark the pre-commit.ci bot's CI as `Required` so that PRs failing the the style checks do not get merged.

Comment (nstarman) [reply .github/workflows/ci_workflows.yml:55]: SGTM.

**Review by Cadair — APPROVED** [review]
:shipit:

**Review by pllim — APPROVED** [review]
We enabled the bot and I updated branch protection rules. Let's get this in!

**Comment by taldcroft** [comment]
Awesome, it worked and fixed a stupid missing-final-newline mistake!! See #13639. Thanks all for getting this working.

**Pull request merged**
